velox/exec/HashProbe.h

```diff
@@ -382,26 +382,71 @@ class HashProbe : public Operator {
     template <typename TOnMiss>
     void advance(vector_size_t row, bool passed, TOnMiss onMiss) {
       if (currentRow != row) {
-        if (currentRow != -1 && !currentRowPassed) {
+        // If this row has been carried over let finishIteration call onMiss
+        // instead.
+        if (currentRow != -1 && !currentRowPassed && currentRow != carriedRow) {
           onMiss(currentRow);
         }
         currentRow = row;
         currentRowPassed = false;
       }
+
       if (passed) {
+        // carriedRow can only be a row that has never passed the filter.  If it
+        // passes there's no need to continue carrying it forward.
+        if (currentRow == carriedRow) {
+          carriedRow = -1;
```

Comment:
Not sure if you want to use std::optional<vector_size_t> for carrierRow to make it clear. Thanks!